cas/ctg_store.c

```diff
@@ -160,82 +213,196 @@ static int ctg_berc(struct m0_ctg_op *ctg_op)
+			  const struct m0_crv *crv)
 {
-	*(uint64_t *)dst = nob;
-	memcpy(dst + M0_CAS_CTG_KV_HDR_SIZE, src, nob);
+	struct generic_value *value = dst->b_addr;
+	M0_PRE(dst->b_nob >= ctg_vbuf_packed_size(src));
+	value->gv_length = src->b_nob;
+	value->gv_version = *crv;
+	memcpy(&value->gv_data, src->b_addr, src->b_nob);
 }
 
-static int ctg_buf_get(struct m0_buf *dst, const struct m0_buf *src)
+/*
+ * Create a new m0_buf filled with CAS-specific data (length).
+ * Note: the function allocates a new buffer.
+ */
+static int ctg_kbuf_get(struct m0_buf *dst, const struct m0_buf *src)
 {
-	m0_bcount_t nob = src->b_nob;
+	struct generic_key *key;
+
+	M0_ENTRY();
 
 	if (M0_FI_ENABLED("cas_alloc_fail"))
 		return M0_ERR(-ENOMEM);
-	dst->b_nob  = src->b_nob + M0_CAS_CTG_KV_HDR_SIZE;
-	dst->b_addr = m0_alloc(dst->b_nob);
-	if (dst->b_addr != NULL) {
-		ctg_memcpy(dst->b_addr, src->b_addr, nob);
+
+	key = m0_alloc(src->b_nob + sizeof(*key));
+	if (key != NULL) {
+		key->gk_length = src->b_nob;
+		memcpy(key->gk_data, src->b_addr, src->b_nob);
+		*dst = M0_BUF_INIT(src->b_nob + sizeof(*key), key);
 		return M0_RC(0);
 	} else
 		return M0_ERR(-ENOMEM);
 }
 
 /**
- * Allocate memory and unpack value having format length + data.
- *
- * @param val destination buffer
- * @param buf source buffer
- * @return 0 if ok else error code
+ * Unpack an an on-disk value data into in-memory format.
+ * The function makes "buf" to point to the user-specific data associated
+ * with the value (see ::generic_value::gv_data).
+ * @param[out] Optional storage for the version of the record.
+ * @return 0 or else -EPROTO if on-disk/on-wire buffer has invalid length.
+ * @see ::ctg_vbuf_pack.
  */
-static int ctg_buf(const struct m0_buf *val, struct m0_buf *buf)
+static int ctg_vbuf_unpack(struct m0_buf *buf, struct m0_crv *crv)
 {
-	int result = -EPROTO;
+	struct generic_value *value;
 
-	M0_CASSERT(sizeof buf->b_nob == 8);
-	if (val->b_nob >= 8) {
-		buf->b_nob = *(uint64_t *)val->b_addr;
-		if (val->b_nob == buf->b_nob + 8) {
-			buf->b_addr = ((char *)val->b_addr) + 8;
-			result = 0;
-		}
-	}
-	if (result != 0)
-		return M0_ERR_INFO(result, "Unexpected: %"PRIx64"/%"PRIx64,
-				   val->b_nob, buf->b_nob);
-	else
-		return M0_RC(result);
+	M0_ENTRY();
+
+	value = buf->b_addr;
+
+	if (buf->b_nob < sizeof(*value))
+		return M0_ERR_INFO(-EPROTO, "%" PRIu64 " < %" PRIu64,
+				   buf->b_nob, sizeof(*value));
+	if (value->gv_length != buf->b_nob - sizeof(*value))
+		return M0_ERR(-EPROTO);
+
+	if (crv != NULL)
+		*crv = value->gv_version;
+
+	buf->b_nob = value->gv_length;
+	buf->b_addr = &value->gv_data[0];
+
+	return M0_RC(0);
+}
+
+/**
+ * Get a m0_cas_ctg pointer from in-memory representation of a meta-index
+ * value.
+ */
+static int ctg_vbuf_as_ctg(const struct m0_buf *buf, struct m0_cas_ctg **ctg)
+{
+	struct meta_value *mv = buf->b_addr - sizeof(struct generic_value);
+
+	M0_ENTRY();
+
+	if (buf->b_nob == sizeof(struct m0_cas_ctg *)) {
+		*ctg = mv->mv_ctg;
```

Comment:
Why don't see just use:
*ctg = (struct m0_cas_ctg *)buf->b_addr;

Reply:
You already have an answer: if we do `*ctg = (struct m0_cas_ctg *)buf->b_addr;` then we are copying the wrong data. b_addr is a double pointer (pointer to pointer to catalogue). It should be something like that:

```c
*ctg = *((struct m0_cas_ctg **) buf->b_addr);
```

Playing with double pointers could be dangerous, so I decided to protect myself and the reader from making such mistakes. I'm using the power of data types to copy the pointer in a relatively safe way. What do you think about that?

Reply:
Yes, agree.
When I did the ctg dump task, I found it is confusing here. 
So, when I read this patch, it is much easier to understand.